File: app/src/main/java/sang/thai/tran/travelcompanion/fragment/RegisterFlightFragment.java
package sang.thai.tran.travelcompanion.fragment;

import android.graphics.Color;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.widget.LinearLayoutCompat;
import android.view.View;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;

import com.nj.imagepicker.ImagePicker;
import com.nj.imagepicker.listener.ImageResultListener;
import com.nj.imagepicker.result.ImageResult;
import com.nj.imagepicker.utils.DialogConfiguration;

import butterknife.BindView;
import butterknife.OnClick;
import butterknife.Optional;
import sang.thai.tran.travelcompanion.R;
import sang.thai.tran.travelcompanion.activity.MainActivity;
import sang.thai.tran.travelcompanion.view.EditTextViewLayout;

import static sang.thai.tran.travelcompanion.activity.MainActivity.NEED_SUPPORT;
import static sang.thai.tran.travelcompanion.utils.AppUtils.openDatePicker;
import static sang.thai.tran.travelcompanion.utils.AppUtils.openTimePicker;

public class RegisterFlightFragment extends BaseFragment {

    @BindView(R.id.email_sign_in_button)
    Button email_sign_in_button;

    @BindView(R.id.et_date)
    EditTextViewLayout et_date;

    @Nullable
    @BindView(R.id.et_hour)
    EditTextViewLayout et_hour;

    @BindView(R.id.et_flight_number)
    EditTextViewLayout et_flight_number;

    @BindView(R.id.et_airline)
    EditTextViewLayout et_airline;

    @BindView(R.id.et_airport_departure)
    EditTextViewLayout et_airport_departure;

    @BindView(R.id.et_arrival_airport)
    EditTextViewLayout et_arrival_airport;

    @BindView(R.id.iv_card_id)
    ImageView iv_card_id;

    @BindView(R.id.tv_card_id)
    TextView tv_card_id;

    @BindView(R.id.iv_flight_ticket_id)
    ImageView iv_flight_ticket_id;

    @BindView(R.id.tv_flight_ticket_id)
    TextView tv_flight_ticket_id;

    public static RegisterFlightFragment newInstance(boolean update) {
        RegisterFlightFragment infoRegisterFragment = new RegisterFlightFragment();
        Bundle bundle = new Bundle();
        bundle.putBoolean(NEED_SUPPORT, update);
        infoRegisterFragment.setArguments(bundle);
        return infoRegisterFragment;
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        et_date.getEditText().setOnFocusChangeListener(new View.OnFocusChangeListener() {
            @Override
            public void onFocusChange(View v, boolean hasFocus) {
                if (hasFocus) {
                    openDepartureDate();
                }
            }
        });

        if (et_hour != null) {
            et_hour.getEditText().setOnFocusChangeListener(new View.OnFocusChangeListener() {
                @Override
                public void onFocusChange(View v, boolean hasFocus) {
                    if (hasFocus) {
                        openDepartureTime();
                    }
                }
            });
        }
    }

    @Override
    protected int layoutId() {
        return R.layout.fragment_register_flight;
    }

    @OnClick(R.id.email_sign_in_button)
    protected void register() {
        ((MainActivity) getActivity()).finishRegister();
    }

    @OnClick(R.id.et_date)
    protected void openDepartureDate() {
        if (getActivity() == null) {
            return;
        }
        openDatePicker(getActivity(), et_date);
    }

    @Optional
    @OnClick(R.id.et_hour)
    protected void openDepartureTime() {
        if (getActivity() == null) {
            return;
        }
        openTimePicker(getActivity(), et_hour);
    }

    @OnClick(R.id.fl_card_id)
    protected void uploadCardId() {
        if (getActivity() == null) {
            return;
        }
        DialogConfiguration configuration = new DialogConfiguration()
                .setTitle("Choose Options")
                .setOptionOrientation(LinearLayoutCompat.VERTICAL)
                .setBackgroundColor(Color.WHITE)
                .setNegativeText("No")
                .setNegativeTextColor(Color.RED)
                .setTitleTextColor(Color.BLUE);

        ImagePicker.build(configuration, new ImageResultListener() {
            @Override
            public void onImageResult(ImageResult imageResult) {
                tv_card_id.setVisibility(View.GONE);
                iv_card_id.setImageBitmap(imageResult.getBitmap());
            }
        }).show(getFragmentManager());
    }

    @OnClick(R.id.fl_flight_ticket_id)
    protected void uploadFlightTicket() {
        if (getActivity() == null) {
            return;
        }
        DialogConfiguration configuration = new DialogConfiguration()
                .setTitle("Choose Options")
                .setOptionOrientation(LinearLayoutCompat.VERTICAL)
                .setBackgroundColor(Color.WHITE)
                .setNegativeText("No")
                .setNegativeTextColor(Color.RED)
                .setTitleTextColor(Color.BLUE);

        ImagePicker.build(configuration, new ImageResultListener() {
            @Override
            public void onImageResult(ImageResult imageResult) {
                iv_flight_ticket_id.setImageBitmap(imageResult.getBitmap());
                tv_flight_ticket_id.setVisibility(View.GONE);
            }
        }).show(getFragmentManager());
    }

}

File: app/src/main/java/sang/thai/tran/travelcompanion/fragment/RegisterFlightFragment.kt
package sang.thai.tran.travelcompanion.fragment

import android.graphics.Color
import android.os.Bundle
import android.view.View
import androidx.appcompat.widget.LinearLayoutCompat
import butterknife.OnClick
import butterknife.Optional
import com.nj.imagepicker.ImagePicker
import com.nj.imagepicker.listener.ImageResultListener
import com.nj.imagepicker.utils.DialogConfiguration
import kotlinx.android.synthetic.main.fragment_register_flight.*
import sang.thai.tran.travelcompanion.R
import sang.thai.tran.travelcompanion.activity.MainActivity
import sang.thai.tran.travelcompanion.model.FlightJobModel
import sang.thai.tran.travelcompanion.model.RegisterModel
import sang.thai.tran.travelcompanion.utils.AppUtils.openDatePicker
import sang.thai.tran.travelcompanion.utils.AppUtils.openTimePicker
import sang.thai.tran.travelcompanion.utils.ApplicationSingleton

open class RegisterFlightFragment : BaseFragment() {

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)

        et_departure_date!!.editText.onFocusChangeListener = View.OnFocusChangeListener { v, hasFocus ->
            if (hasFocus) {
                openDepartureDate()
            }
        }
        et_departure_date.setOnClickListener { openDepartureDate() }

        if (et_departure_hour != null) {
            et_departure_hour!!.editText.onFocusChangeListener = View.OnFocusChangeListener { v, hasFocus ->
                if (hasFocus) {
                    openDepartureTime()
                }
            }
            et_departure_hour.setOnClickListener { openDepartureTime() }
        }
        email_sign_in_button.setOnClickListener {
            //            (activity as MainActivity).finishRegister()
//            registerApi()
            ApplicationSingleton.getInstance().userInfo.flightJobModel = createFlightJobModel()
            (activity as MainActivity).onBackPressed()
        }
        if (fl_card_id != null) {
            fl_card_id.setOnClickListener { uploadCardId() }
        }
        if (fl_flight_ticket_id != null)
            fl_flight_ticket_id.setOnClickListener { uploadFlightTicket() }
    }

    override fun layoutId(): Int {
        return R.layout.fragment_register_flight
    }

    private fun openDepartureDate() {
        if (activity == null) {
            return
        }
        openDatePicker(activity, et_departure_date)
    }

    @Optional
    @OnClick(R.id.et_departure_hour)
    fun openDepartureTime() {
        if (activity == null) {
            return
        }
        openTimePicker(activity, et_departure_hour)
    }

    @Optional
    @OnClick(R.id.fl_card_id)
    fun uploadCardId() {
        if (activity == null) {
            return
        }
        val configuration = DialogConfiguration()
                .setTitle("Choose Options")
                .setOptionOrientation(LinearLayoutCompat.VERTICAL)
                .setBackgroundColor(Color.WHITE)
                .setNegativeText("No")
                .setNegativeTextColor(Color.RED)
                .setTitleTextColor(Color.BLUE)

        ImagePicker.build(configuration, ImageResultListener { imageResult ->
            if (tv_card_id != null) {
                tv_card_id!!.visibility = View.GONE
            }
            if (iv_card_id != null) {
                iv_card_id!!.setImageBitmap(imageResult.bitmap)
            }
        }).show(fragmentManager!!)
    }

    @Optional
    @OnClick(R.id.fl_flight_ticket_id)
    fun uploadFlightTicket() {
        if (activity == null) {
            return
        }
        val configuration = DialogConfiguration()
                .setTitle("Choose Options")
                .setOptionOrientation(LinearLayoutCompat.VERTICAL)
                .setBackgroundColor(Color.WHITE)
                .setNegativeText("No")
                .setNegativeTextColor(Color.RED)
                .setTitleTextColor(Color.BLUE)

        ImagePicker.build(configuration, ImageResultListener { imageResult ->
            if (iv_flight_ticket_id != null) {
                iv_flight_ticket_id!!.setImageBitmap(imageResult.bitmap)
            }
            if (tv_flight_ticket_id != null) {
                tv_flight_ticket_id!!.visibility = View.GONE
            }
        }).show(fragmentManager!!)
    }

    companion object {

        fun newInstance(update: Boolean): RegisterFlightFragment {
            val infoRegisterFragment = RegisterFlightFragment()
            val bundle = Bundle()
            bundle.putBoolean(MainActivity.NEED_SUPPORT, update)
            infoRegisterFragment.arguments = bundle
            return infoRegisterFragment
        }
    }

    override fun createRegisterFlight(): RegisterModel {
        var registerModel = ApplicationSingleton.getInstance().registerModel
        if (registerModel == null) {
            registerModel = RegisterModel()
        }

        registerModel.departureDateFrom = et_departure_date?.text.toString() + " " + et_departure_hour?.text.toString()
        registerModel.id = ApplicationSingleton.getInstance().userInfo.code
        registerModel.airlineOption = et_airline?.text
        registerModel.flightNumber = et_flight_number?.text
        registerModel.departureAirport = et_airport_departure?.text
        registerModel.arrivalAirport = et_arrival_airport?.text
        addMoreService(registerModel)
        return registerModel
    }

    private fun createFlightJobModel() : FlightJobModel {
        val registerModel = FlightJobModel()
        registerModel.ticketDeprtDate = et_departure_date?.text.toString() + " " + et_departure_hour?.text.toString()
        registerModel.airline = et_airline?.text
        registerModel.flightNumber = et_flight_number?.text
        registerModel.ticketDeprtAirport = et_airport_departure?.text
        registerModel.ticketArrAirport = et_arrival_airport?.text
        registerModel.ticketImage = "http"
        return registerModel
    }

    protected open fun addMoreService(registerModel: RegisterModel) {

    }
}
